tests/test_api.py

```diff
@@ -0,0 +1,386 @@
+handling upload-related functionality).
+"""
+from unittest.mock import MagicMock
+from datetime import datetime
+
+from flask import _request_ctx_stack
+
+import pytest
+from cidc_api.models import (
+    Users,
+    DownloadableFiles,
+    Permissions,
+    TrialMetadata,
+    UploadJobs,
+    UploadJobStatus,
+    CIDCRole,
+)
+
+TEST_RECORD_ID = 1
+new_user = {"email": "test-admin@email.com"}
+user = {
+    **new_user,
+    "id": TEST_RECORD_ID,
+    "role": CIDCRole.ADMIN.value,
+    "approval_date": datetime.now(),
+}
+trial_metadata = {"id": TEST_RECORD_ID, "trial_id": "foo", "metadata_json": {}}
+downloadable_file = {
+    "id": TEST_RECORD_ID,
+    "trial_id": trial_metadata["trial_id"],
+    "file_name": "",
+    "upload_type": "olink",
+    "data_format": "",
+    "object_url": "",
+    "uploaded_timestamp": datetime.now(),
+    "file_size_bytes": 1,
+}
+permission = {
+    "id": TEST_RECORD_ID,
+    "granted_to_user": TEST_RECORD_ID,
+    "trial_id": trial_metadata["trial_id"],
+    "upload_type": downloadable_file["upload_type"],
+}
+upload_job = {
+    "id": TEST_RECORD_ID,
+    "trial_id": trial_metadata["trial_id"],
+    "uploader_email": user["email"],
+    "upload_type": downloadable_file["upload_type"],
+    "metadata_patch": {},
+    "gcs_xlsx_uri": "",
+    "multifile": False,
+    "status": UploadJobStatus.STARTED.value,
+}
+
+
+@pytest.fixture
+def app_with_admin_user(app, monkeypatch):
+    def fake_auth(*args):
+        _request_ctx_stack.top.current_user = Users(**user)
+        return True
+
+    monkeypatch.setattr(app.auth, "authorized", fake_auth)
+
+    return app
+
+
+ETAG = "test-etag"
+
+
+@pytest.fixture
+def db_with_records(db):
+    extra = {"_etag": ETAG}
+    db.add(Users(**user, **extra))
+    db.add(TrialMetadata(**trial_metadata, **extra))
+    db.commit()
+
+    db.add(DownloadableFiles(**downloadable_file, **extra))
+    db.add(Permissions(**permission, **extra))
+    db.add(UploadJobs(**upload_job, **extra))
+    db.commit()
+
+    return db
+
+
+# Configuration for resource tests below. For each resource, the following keywords are supported:
+#   `json` (required): a JSON instance of this resource.
+#   `model` (required): the SQLAlchemy model for this resource.
+#   `allowed_methods` (required): the HTTP methods this resource supports.
+#   `POST_setup`: a list of other resources to add to the database before POSTing this resource.
+#   `PATCH_json` (required if "PATCH" in `allowed_methods`): a JSON patch update for this resource.
+#   `filters`: a dictionary containing two entries representing possible filter queries:
+#       `empty`: a query filter that should return empty results.
+#       `one`: a query filter that should return exactly one result.
+#   `additional_records`: a list of JSON instances of this resource to insert before testing pagination.
+#   `mocks`: a list of functions that accept pytest's `monkeypatch` as their argument.
+resource_requests = {
```

Comment:
I would think on somehow splitting that huge object that represents not only multiple tests, but multiple test for multiple models, which is too much I think.

That would also allow to include `json` instance of a resource directly into every such record, brining them together as something like
```
resource_requests = {}
resource_requests['new_users']={
        "json": {
            **new_user,
            "id": TEST_RECORD_ID,
            "role": CIDCRole.ADMIN.value,
            "approval_date": datetime.now(),
        },
        "model": Users,
        "allowed_methods": {"POST", "PATCH", "GET"},
        "PATCH_json": {"role": CIDCRole.CIMAC_USER.value},
...
```